### PR TITLE
Editor: Cleanup some includes dependencies

### PR DESCRIPTION
**Pull request by akien-mga**
Follow-up to #57641, CC @Geometror.
Follow-up to #57306, CC @trollodel.

Removes some unnecessary includes from `editor_node.h`, and instead add
those where they're used.

Removes unnecessary `editor_node.h` includes in various editor classes.

Renames `dynamicfont` to `dynamic_font` in a couple files.

Misc cleanup while jumping through that rabbit hole.

---

Incremental rebuild time after cosmetic change in `editor/editor_node.h` (note: laptop on battery):

- Before this PR: 4m 37s
- After this PR: 3m 59s (**14% speedup**)

### CODE REVIEW

**Comment by akien-mga** [comment]
For the record, in some of those changes I could have used more forward declares (e.g. in `editor/action_map_editor.h` with all the new `scene/gui` includes needed after the removal of `editor/editor_data.h`). But we don't have a clear policy yet on when to use forward declares or not - and I didn't want to go too deep that rabbit hole in that PR specifically.

I don't know if we should use forward declares systematically for all headers. On the other hand, I think we could have a policy that any header included in e.g. more than 3-5 files (recursively) *should* use forward declares as that's where the gains will be most important.

**Comment by trollodel** [comment]
> I don't know if we should use forward declares systematically for all headers. 
In most cases, I see that plugins are included only by EditorNode. While I agree that EditorNode is a giant class to compile, I think it's possible to move plugin registration to a dedicated file, like other register_*.cpp files in the codebase, or like editor plugins registered in modules.

File: editor/connections_dialog.h [hunk 1]
```diff
@@ -35,16 +35,20 @@
...
 #include "scene/gui/line_edit.h"
 #include "scene/gui/menu_button.h"
+#include "scene/gui/option_button.h"
 #include "scene/gui/popup.h"
+#include "scene/gui/popup_menu.h"
+#include "scene/gui/spin_box.h"
 #include "scene/gui/tree.h"
 
-class PopupMenu;
+class EditorNode;
```
Comment (trollodel) [inline editor/connections_dialog.h:50]: It's not used.

File: modules/csg/csg_gizmos.h [hunk 2]
```diff
@@ -35,6 +35,8 @@
 #include "editor/editor_plugin.h"
 #include "editor/plugins/node_3d_editor_gizmos.h"
 
+class EditorNode;
```
Comment (trollodel) [inline modules/csg/csg_gizmos.h:38]: It's not used.

**Review by trollodel — APPROVED** [review]
Did a quick pass to the changes. As you said, there are more classes that can be forward declared, but this PR it's already a great work. 
I'm working to another PR that removes, as a side effect, other includes not removed here.

File: editor/connections_dialog.h (hunk 1 above)
Comment (akien-mga) [reply editor/connections_dialog.h:50]: Right, I started this work before #57306 was merged so I had added a few forward declares, seems I didn't remove them all.

**Comment by pycbouh** [comment]
> I think we could have a policy that any header included in e.g. more than 3-5 files (recursively) should use forward declares as that's where the gains will be most important.

This will be hard to track when making or reviewing a PR. You typically don't go through all includes to see if they are included elsewhere and how many times. So to maintain this we'd have to either keep a good mind map of yet unused but getting there includes, or do cleanup passes every so often.

File: editor/connections_dialog.h (hunk 1 above)
Comment (akien-mga) [reply editor/connections_dialog.h]: Fixed, together with a handful of other unneeded `EditorNode` forward declares I had missed.

**Comment by akien-mga** [comment]
> > I think we could have a policy that any header included in e.g. more than 3-5 files (recursively) should use forward declares as that's where the gains will be most important.
> 
> This will be hard to track when making or reviewing a PR. You typically don't go through all includes to see if they are included elsewhere and how many times. So to maintain this we'd have to either keep a good mind map of yet unused but getting there includes, or do cleanup passes every so often.

That's true, though we could have folder specific guidelines. As pointed out in https://github.com/godotengine/godot/pull/58136#issuecomment-1040084559 for example, editor plugins don't really need to be forward declared, as they're usually only included in `editor_node.cpp` (with some exceptions).

On the other hand, `core`, `scene` and `servers` stuff should likely use forward declares as much as possible as those are included everywhere.

I did a quick test reviewing a `scene/gui` class to add forward declares and I think it already raises a number of questions: https://github.com/akien-mga/godot/commit/7a37a59c86285960de1d556087661afcfb0c90fb

- Do we forward declare *everything* that can be?
- Do we forward declare classes which are already being included in some of the other non-forward declared dependencies? (E.g. stuff included by `scene/gui/control.h` like `Size2`, `Ref<T>`, `String`, `StringName`, `ImageTexture` and `Texture2D`)
  * Or do we restrict this to forward declaring only non-core stuff, so from the above list only `ImageTexture` and `Texture2D` (because they're included by Control -> CanvasItem -> CanvasItemMaterial -> Material -> Texture and it's hard to have a mental model of this at all times to decide).

**Review by Geometror — APPROVED** [review]
Nice work :)
I would suggest using forward declarations whenever possible but exclude core types (and maybe resource types [scene/resources/*]) like those you mentioned above. If they aren’t already covered with that, I would also exclude all template classes since forward declaring them takes up quite some space and is rather ugly. 
Heavy use of forward declarations around GUI nodes (especially in the dialog classes) might be worth it, as it is straight-forward to implement (in most cases the header just contains pointers of these types) and breaks up many critical dependency chains, but this should be done in a separate PR.

**Pull request merged**